src/GitVersionTask/GenerateGitVersionInformation.cs

```diff
@@ -1,42 +1,99 @@
+            Output output = null;
+            try
+            {
+                output = InnerExecute(input );
+            }
+            catch (WarningException errorException)
+            {
+                logger.LogWarning(errorException.Message);
+                output = new Output();
+            }
+            catch (Exception exception)
+            {
+                logger.LogError("Error occurred: " + exception);
+                throw;
+            }
+            finally
+            {
+                Logger.Reset();
+            }
 
-        protected override void InnerExecute()
+            return output;
+        }
+
+        private static Output InnerExecute( Input input )
         {
-            if (GetVersionVariables(out var versionVariables)) return;
+            var execute = GitVersionTaskBase.CreateExecuteCore();
+            if (!execute.TryGetVersion(input.SolutionDirectory, out var versionVariables, input.NoFetch, new Authentication()))
+            {
+                return null;
+            }
 
-            var fileExtension = TaskUtils.GetFileExtension(Language);
-            var fileName = $"GitVersionInformation.g.{fileExtension}";
+            var fileWriteInfo = input.IntermediateOutputPath.GetWorkingDirectoryAndFileNameAndExtension(
+                input.Language,
+                input.ProjectFile,
+                ( pf, ext ) => $"GitVersionInformation.g.{ext}",
+                ( pf, ext ) => $"GitVersionInformation_{Path.GetFileNameWithoutExtension( pf )}_{Path.GetRandomFileName()}.g.{ext}"
+                );
 
-            if (IntermediateOutputPath == null)
+            var output = new Output()
             {
-                fileName = $"GitVersionInformation_{Path.GetFileNameWithoutExtension(ProjectFile)}_{Path.GetRandomFileName()}.g.{fileExtension}";
-            }
+                GitVersionInformationFilePath = Path.Combine( fileWriteInfo.WorkingDirectory, fileWriteInfo.FileName )
+            };
+            var generator = new GitVersionInformationGenerator( fileWriteInfo.FileName, fileWriteInfo.WorkingDirectory, versionVariables, new FileSystem());
+            generator.Generate();
 
-            var workingDirectory = IntermediateOutputPath ?? TempFileTracker.TempPath;
+            return output;
+        }
 
-            GitVersionInformationFilePath = Path.Combine(workingDirectory, fileName);
+        public sealed class Input
+        {
+            public string SolutionDirectory { get; set; }
 
-            var generator = new GitVersionInformationGenerator(fileName, workingDirectory, versionVariables, new FileSystem());
-            generator.Generate();
+            public string ProjectFile { get; set; }
+
+            public string IntermediateOutputPath { get; set; }
+
+            public string Language { get; set; }
+
+            public bool NoFetch { get; set; }
+        }
+
+        private static Boolean ValidateInput( this Input input )
```

Comment:
@stazz, why isn't this method placed inside the `Input` class itself?

Reply:
No particular reason. I guess the validity input depends a bit on a context (which I admit in this case is exactly one). Can move it as instance method of `Input` class on Sunday.